src/adagrams.js

```diff
@@ -1,15 +1,162 @@
+    all_letters[j] = all_letters[i]; // swapping begins here
+    all_letters[i] = temp;
+  }
+
+  const handFunction = function (all_letters) {
+    let hand = [];
+    for (const letter of all_letters) {
+      while (hand.length < 10) {
+        hand.push(letter);
+        break;
+      }
+    }
+    return hand;
+  };
+  return handFunction(all_letters);
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
   // Implement this method for wave 2
+  for (let i = 0; i < input.length; i++) {
+    let letter = input[i];
+    let letterIndex = lettersInHand.indexOf(letter); // indexOf letter inside lettersInHand
+    if (letterIndex < 0) return false;
+    lettersInHand.splice(letterIndex, 1); // (starting index, how many to remove starting from starting index)
+  }
+  return true;
 };
 
 export const scoreWord = (word) => {
   // Implement this method for wave 3
+  const scoreChart = {
+    1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+    2: ["D", "G"],
+    3: ["B", "C", "M", "P"],
+    4: ["F", "H", "V", "W", "Y"],
+    5: ["K"],
+    8: ["J", "X"],
+    10: ["Q", "Z"],
+  };
+
+  let score = 0;
+  const scoreLetters = Object.values(scoreChart); // seperates into array of arrays of letters
+  const scorePoints = Object.keys(scoreChart); // seperates into array of numbers
+
+  for (let i = 0; i < word.length; i++) {
+    let currentLetter = word[i].toUpperCase();
+    for (let j = 0; j < scoreLetters.length; j++) {
+      if (scoreLetters[j].includes(currentLetter)) {
+        score = score + parseInt(scorePoints[j]); // scorePoints[j] has same index of scoreLetters[j]
+      }
+    }
+  }
+  if (
+    word.length === 7 ||
+    word.length === 8 ||
+    word.length === 9 ||
+    word.length === 10
+  ) {
+    score = score + 8;
+  }
+  return score;
 };
 
 export const highestScoreFrom = (words) => {
   // Implement this method for wave 4
+  const scoreChart = {
+    1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+    2: ["D", "G"],
+    3: ["B", "C", "M", "P"],
+    4: ["F", "H", "V", "W", "Y"],
+    5: ["K"],
+    8: ["J", "X"],
+    10: ["Q", "Z"],
+  };
+
+  const scoreLetters = Object.values(scoreChart);
+  const scorePoints = Object.keys(scoreChart);
+
+  let highScore = 0;
+  let highScoreWord = "";
+
+  for (let word of words) {
+    let score = 0;
+    for (let i = 0; i < word.length; i++) {
+      let currentLetter = word[i].toUpperCase();
+      for (let j = 0; j < scoreLetters.length; j++) {
+        if (scoreLetters[j].includes(currentLetter)) {
+          score = score + parseInt(scorePoints[j]);
+        }
+      }
+    }
+    if (
+      word.length === 7 ||
+      word.length === 8 ||
+      word.length === 9 ||
+      word.length === 10
+    ) {
+      score = score + 8;
+    }
+    if (highScore < score) {
+      highScore = score;
+      highScoreWord = word;
+    } else if (highScore === score) {
+      if (word.length < 10) {
+        if (highScoreWord.length === 10) {
+          break;
```

Comment:
This is unreachable code because the conditional in L146 checks that the length of `word` is less than 10 to enter this block.